### PR TITLE
feat(files_sharing): Improve user experience when user attempts to download folder with non-downloadable files

### PR DESCRIPTION
**Pull request by Koc**
This PR is continuation of the recently merged #57573. 

Current `master` branch requires to perform recursive iteration twice:
- 1st time when we check can node to be downloaded
- 2nd time when we actually populate zip archive

With this PR we recursively iterating only once as checks performed right before node added to archive.

But that's not all. Current `master` branch not allows download anything if downloaded folder contains at least 1 non-downloadable file. With this PR non-downloadable files are just excluded from the archive. At the same downloadable files added to it.

## TODO
- [x] Check more edge-cases (e.g. shared and own files located in nested folder)
- [ ] Actualize tests

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)
- [ ] [Labels added](https://github.com/nextcloud/server/labels) where applicable (ex: bug/enhancement, `3. to review`, feature component)
- [ ] [Milestone added](https://github.com/nextcloud/server/milestones) for target branch/version (ex: 32.x for `stable32`)

### CODE REVIEW

File: apps/files_sharing/lib/Listener/BeforeZipCreatedListener.php [hunk 1]
```diff
@@ -32,28 +32,38 @@ public function handle(Event $event): void {
 			return;
 		}
 
-		$dir = $event->getDirectory();
-		$files = $event->getFiles();
+		$user = $this->userSession->getUser();
+		if (!$user) {
+			// fixme: do we need check anything for anonymous users?
+			$event->setSuccessful(true);
```
Comment (come-nc) [inline apps/files_sharing/lib/Listener/BeforeZipCreatedListener.php]: a return is missing here I believe

**Review by come-nc — COMMENTED** [review]
I do not understand the PR, it checks the same files twice, once as paths, once as nodes?
Why is that needed? Why was it not working before?

**Comment by Koc** [comment]
@come-nc 
> Why was it not working before?

Good question, but I don't know. For some reason we're still able to download non-downloadable files using workarounds that I explained in the PR description. I've double-tested it on `master` branch before PR opening.

Regarding duplicated checks - I've added one more commit that simplifies this. So, now we recursively collecting files to download only once and perform check only once. Also in previous implementation non-downloaded files break whole download (check not works, but idea was the same). Now we're just skipping such files and not break download process.

**Comment by Koc** [comment]
@susnux @AndyScherzinger guys, can you take a look on it, please?

**Comment by AndyScherzinger** [comment]
@Koc I pinge @sorbaugh to have engineering reviewing the PR. I am not a php/js dev unfortunately.

File: apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php [hunk 2]
```diff
@@ -137,7 +145,20 @@ public function handleDownload(Request $request, Response $response): ?bool {
...
+		$rootNodes = empty($files) ? $folder->getDirectoryListing() : [];
+		foreach ($files as $path) {
+			$child = $node->getChild($path);
+			assert($child instanceof Node);
+			$rootNodes[] = $child->getNode();
+		}
+		$allNodes = [];
+		foreach ($rootNodes as $rootNode) {
+			foreach ($this->iterateNodes($rootNode) as $node) {
+				$allNodes[] = $node;
```
Comment (juliusknorr) [inline apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php]: If we create such an array and pass it along I could imagine we run into memory limits here with very large directory structures. Would it make sense to rather pass the generator itself?

**Review by juliusknorr — COMMENTED** [review]
Code wise this makes sense and solves the issue 👍 

Just one comment left

File: apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php (hunk 2 above)
Comment (Koc) [reply apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php]: yeah, I thought about that too. Refactored, I hope laziness will work, please check

**Comment by Koc** [comment]
@juliusknorr I've updated PR description. Actually fix itself done in mentioned PR. But this PR improves behavior

**Comment by susnux** [comment]
I am not sure about this:

> With this PR non-downloadable files are just excluded from the archive.

This seems very unexpected to me:
- I try to download a folder.
- Download **succeeds** 
- I extract it and some files **silently** are missing

e.g. imagine you "backup" the files and then delete them, now you lost access to some of them.

With the explicit error we have currently you at least get notified there are files you cannot download.
So best to check UX here with @nextcloud/designers

**Comment by nickvergessen** [comment]
> With the explicit error we have currently you at least get notified there are files you cannot download.
So best to check UX here

We could run a pre-check and ask the user which one is better?
Because otherwise I can simply share a file with you and you can no longer download your Shared/ folder?

**Comment by come-nc** [comment]
I had a similar question related to encrypted files, if a file is corrupted some users would like to still be able to download the rest of the folder. But as far as I know it’s not possible to both create and return a zip and show errors? So to me it feels too dangerous to return incomplete data silently.

**Comment by susnux** [comment]
One option I think we could have:
- try to download
- it fails like it is currently
- UI shows the error with option to download anyways
- we retry the download with a `forceIgnoreErrors=true` option (or any other name)

But thats why I would like to have designers opinion on the UX part.

**Comment by salmart-dev** [comment]
> One option I think we could have:
> 
>     * try to download
> 
>     * it fails like it is currently
> 
>     * UI shows the error with option to download anyways
> 
>     * we retry the download with a `forceIgnoreErrors=true` option (or any other name)
> 
> 
> But thats why I would like to have designers opinion on the UX part.

IMO Since the plugin is used in more than just Nextcloud clients, it would be best if we could implement a solution that doesn't depend on the client doing something specific so that users on any client have the information that something is wrong with the download, or at least a combination of both.

I was looking at the plugin to understand the cause of a similar problem and I ended up drafting a solution that would help users, server admins and our support to understand why files are missing and, most importantly for the user, which files: create the archive and at the end add a file `ARCHIVE_MISSING_FILES.{txt|json|csv|...}` if any files are missing, detailing the reason why. Some already known reasons can be: `File locked`, `Storage not available`, `File not found`, `Read XXX bytes but expected to read YYY from storage` and so on.

This would be independent from the cllient but of course would require the user to actually check the archive to see if any of the important files are missing, and if so why.

**Comment by susnux** [comment]
> This would be independent from the cllient but of course would require the user to actually check the archive to see if any of the important files are missing, and if so why.

But it would still hide it in the first place.

IMHO if the download does not work as requested by the user it must fail hard with appropriate error code (e.g. 423 locked or 403 forbidden). 
This ensures you not silently hide errors which lead to data corruption (e.g. missing expected content).

Thats why I think a second request with a "ok I am aware of this, please continue anyways" header is the best option here for user experience.
If you use this outside of Nextcloud then, I think it is still valid to fail initially.

**Comment by salmart-dev** [comment]
> IMHO if the download does not work as requested by the user it must fail hard with appropriate error code (e.g. 423 locked or 403 forbidden).

But this cannot be done once we begin streaming the archive because we don't use `Content-Length` so the browser doesn't know the transfer stopped because of an error.

The only way to reliably implement this behaviour would be to build the archive on the server and then serve it to the user when ready.

**Comment by kra-mo** [comment]
I don't have much more to add, I agree with @susnux on failure from the design side.

**Comment by kra-mo** [comment]
If however, the technical limitations really are that it cannot be done that way, would it be possible to give the archive a different name? Adding an ` (Incomplete)` suffix or something like that, to make sure that the user can see upfront that not all files are present.

**Comment by salmart-dev** [comment]
> If however, the technical limitations really are that it cannot be done that way, would it be possible to give the archive a different name? Adding an ` (Incomplete)` suffix or something like that, to make sure that the user can see upfront that not all files are present.

It can only be done when we know that some files or directories are inaccessible, like for the case of shares that are not downloadable. It cannot be done for any file/folder missing or incomplete due to temporary network issues or storage issues unfortunately.

**Comment by kra-mo** [comment]
I see. But actually, when issues are temporary, I feel like hard failure might be a better option anyway? Since then it is highly likely the user would want to retry the operation.

File: apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php [hunk 3]
```diff
@@ -169,13 +187,13 @@ public function handleDownload(Request $request, Response $response): ?bool {
 			$rootPath = dirname($folder->getPath());
 		}
 
-		$streamer = new Streamer($tarRequest, -1, count($content), $this->timezoneFactory);
+		$streamer = new Streamer($tarRequest, -1, count($rootNodes), $this->timezoneFactory);
```
Comment (salmart-dev) [inline apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:190]: Spotted another issue that is present in the current implementation: the count of files is not accounting for files in subfolders. If the total number of files is greater than 65535 this could cause issues, since we'd need to use the headers for zip64.

**Comment by salmart-dev** [comment]
> I see. But actually, when issues are temporary, I feel like hard failure might be a better option anyway? Since then it is highly likely the user would want to retry the operation.

This is not possible to do with the current implementation as abruptly stopping the download will result in a corrupted archive like now. The only way to implement this reliably would be to generate the archive on the server and then informing the client whether the archive is ready or if not, inform it of the errors. But this also means that if the user tries to download 1TB of data, that the server will need 1TB of available disk space to write the archive, which defeats the purpose of the plugin.

File: apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php (hunk 3 above)
Comment (salmart-dev) [reply apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php:190]: I realised size=-1 forces the zip64 when zip is used

**Comment by susnux** [comment]
> This is not possible to do with the current implementation as abruptly stopping the download will result in a corrupted archive like now. The only way to implement this reliably would be to generate the archive on the server and then informing the client whether the archive is ready or if not, inform it of the errors. But this also means that if the user tries to download 1TB of data, that the server will need 1TB of available disk space to write the archive, which defeats the purpose of the plugin.

But this PR already scans all files before streaming, so we can properly return HTTP 400 when a file does not have download permissions.

For runtime errors like `locked` this is not possible, true, but for locked files I am not sure if its not best to skip them and retry later to add them (with some max. time).
